### PR TITLE
Destroy Associations in Background Job

### PR DESCRIPTION
**Pull request by rafaelfranca**
Updated version of https://github.com/rails/rails/pull/36912

cc @gwincr11

### CODE REVIEW

File: activerecord/lib/active_record/associations.rb [hunk 1]
```diff
@@ -1357,6 +1358,9 @@ module ClassMethods
         #   association objects.
         # [:strict_loading]
         #   Enforces strict loading every time the associated record is loaded through this association.
+        # [:owner_ensuring_destroy]
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/associations.rb:1361]: I'm not happy with this option name. Any suggestions?

File: activerecord/test/schema/schema.rb [hunk 2]
```diff
@@ -288,6 +291,50 @@
     t.string :name
   end
 
+  create_table :destroy_later_parents, force: true, id: false do |t|
```
Comment (rafaelfranca) [inline activerecord/test/schema/schema.rb:294]: I'd love to delete those new tables and reuse existing tables but I need to understand why those models are needed now.

File: activerecord/lib/active_record/associations.rb (hunk 1 above)
Comment (jonathanhefner) [reply activerecord/lib/active_record/associations.rb:1361]: Do we have a naming convention for options that refer to method names?  If not, is one of these preferable:

* `ensure_owner_destroyed: :deleted?`
* `ensures_owner_destroyed: :deleted?`
* `ensure_owner_destroyed_with: :deleted?`

Also, would `verify` be better than `ensure`, since `ensure` sounds like it will make it happen if it is not already done?

* `verify_owner_destroyed_with: :deleted?`

Comment (rafaelfranca) [reply activerecord/lib/active_record/associations.rb:1361]: `verify_owner_destroyed_with` is a good name. Thank you for the suggestion @jonathanhefner

File: activerecord/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,21 @@
...
+
+      # Destroy exports 30 days after completing them
+      destroy_later after: 30.days, if: -> { status_previously_changed? && completed? }
+
+      # Destroy exports 30 days after completing them, ensuring they're still completed at destroy time
+      destroy_later after: 30.days, if: -> { status_previously_changed? && completed? }, ensuring: :completed?
+    end
+    ```
+
+    *DHH*, *George Claghorn, Cory Gwin*
```
Comment (eileencodes) [inline activerecord/CHANGELOG.md:17]: There's stars missing around the last 2 names.

```suggestion
    *DHH*, *George Claghorn*, *Cory Gwin*
```

File: activerecord/lib/active_record/associations.rb [hunk 4]
```diff
@@ -1291,6 +1291,7 @@ module ClassMethods
         #
         #   * <tt>nil</tt> do nothing (default).
         #   * <tt>:destroy</tt> causes all the associated objects to also be destroyed.
+        #   * <tt>:destroy_later</tt> causes all the associated object to be destroyed in a background job.
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations.rb:1294]:
```suggestion
        #   * <tt>:destroy_later</tt> destroys all the associated objects in a background job.
```

File: activerecord/lib/active_record/associations.rb [hunk 5]
```diff
@@ -1357,6 +1358,9 @@ module ClassMethods
         #   association objects.
         # [:strict_loading]
         #   Enforces strict loading every time the associated record is loaded through this association.
+        # [:owner_ensuring_destroy]
+        #   Specifies a instance method name that will be called in the owner when executing the background job to
+        #   delete the associated record to determine if the record should be deleted.
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations.rb:1363]: This sentence is hard to parse. How about something like `Specifies an instance method that will be called by the owner to destroy the associated record. The associated records will be destroyed in a background job.`

File: activerecord/lib/active_record/associations.rb [hunk 6]
```diff
@@ -1439,6 +1443,7 @@ def has_many(name, scope = nil, **options, &extension)
         #
         #   * <tt>nil</tt> do nothing (default).
         #   * <tt>:destroy</tt> causes the associated object to also be destroyed
+        #   * <tt>:destroy_later</tt> causes all the associated object to be destroyed in a background job.
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations.rb:1446]:
```suggestion
        #   * <tt>:destroy_later</tt> destroys all the associated objects in a background job.
```

File: activerecord/lib/active_record/associations.rb [hunk 7]
```diff
@@ -1499,6 +1504,9 @@ def has_many(name, scope = nil, **options, &extension)
         #   +:inverse_of+ to avoid an extra query during validation.
         # [:strict_loading]
         #   Enforces strict loading every time the associated record is loaded through this association.
+        # [:owner_ensuring_destroy]
+        #   Specifies a instance method name that will be called in the owner when executing the background job to
+        #   delete the associated record to determine if the record should be deleted.
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations.rb:1509]: Same suggestion as above: `Specifies an instance method that will be called by the owner to destroy the associated record. The associated records will be destroyed in a background job.`

File: activerecord/lib/active_record/associations.rb [hunk 8]
```diff
@@ -1647,6 +1656,9 @@ def has_one(name, scope = nil, **options)
         #   be initialized with a particular record before validation.
         # [:strict_loading]
         #   Enforces strict loading every time the associated record is loaded through this association.
+        # [:owner_ensuring_destroy]
+        #   Specifies a instance method name that will be called in the owner when executing the background job to
+        #   delete the associated record to determine if the record should be deleted.
```
Comment (eileencodes) [inline activerecord/lib/active_record/associations.rb:1661]: Same update: `Specifies an instance method that will be called by the owner to destroy the associated record. The associated records will be destroyed in a background job`

**Review by eileencodes — CHANGES_REQUESTED** [review]
Looking good, I have some comments on docs though.

File: activerecord/lib/active_record/associations.rb (hunk 5 above)
Comment (jonathanhefner) [reply activerecord/lib/active_record/associations.rb:1363]: I think the intent may be more like `Specifies an instance method to verify the owner is destroyed when destroying the associated object in a background job.`

File: activerecord/lib/active_record/associations.rb (hunk 1 above)
Comment (kaspth) [reply activerecord/lib/active_record/associations.rb:1361]: > Also, would `verify` be better than `ensure`, since `ensure` sounds like it will make it happen if it is not already done?

`ensure` is meant as a prerequisite, like in `destroy_later ensuring: :due?`. I read it as "can't proceed if this hasn't happened".

And I do think there's something nice about reusing the wording, though perhaps it's also confusing because it's targeting the owner and not the associated records.

My take would be:

```ruby
has_many :comments, dependent: :destroy_later, ensuring_owner_was: :destroyed?
```

`_was` to imply the check is run in the job i.e. later, `ensuring` to match the prerequisite option on `record.destroy_later` also has, and `owner` to indicate the check is on the owner of course. A lot of the meaning is filled in via the supplied predicate granted, but that seems ok for the option to me.

File: activerecord/lib/active_record/associations/builder/association.rb [hunk 9]
```diff
@@ -120,7 +120,11 @@ def self.valid_dependent_options
       raise NotImplementedError
     end
 
-    def self.check_dependent_options(dependent)
+    def self.check_dependent_options(dependent, model)
+      if dependent == :destroy_later && !model.destroy_association_later_job
+        err_message = "ActiveJob is required to use destroy_later on dependencies"
```
Comment (kaspth) [inline activerecord/lib/active_record/associations/builder/association.rb:125]: dependencies => associations?

**Review by kaspth — APPROVED** [review]
Some small suggestions, but really cool we're getting this!

I could use `destroy_later` for something I'm working on right now too 😄

File: activerecord/test/schema/schema.rb (hunk 2 above)
Comment (gwincr11) [reply activerecord/test/schema/schema.rb:294]: I added new models because adding the destroy later arguments to has many relations on other models was causing issues with other tests and I could not figure out a way to add it cleanly for just the tests it affects. It also felt much cleaner to isolate the tests.

Comment (gwincr11) [reply activerecord/test/schema/schema.rb:294]: Also there were a lot of edge cases to test against 😅

File: activerecord/CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,21 @@
...
+
```
Comment (gwincr11) [inline activerecord/CHANGELOG.md:18]: I did not add the has many destroy later into the change log should I? Destroy later is a separate feature.

File: activerecord/lib/active_record/destroy_later.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        after_commit -> { destroy_later after: after, ensuring: ensuring },
+          on: option_if ? %i[ create update ] : :create, if: option_if
+      end
+    end
+
+    # Schedules a record to be destroyed after a specified amount of time.
+    #
+    # Optionally pass the name of an instance method in +ensuring:+. If the given method returns false at the
+    # scheduled destroy time, the record will not be destroyed.
+    def destroy_later(after:, ensuring: nil)
```
Comment (kaspth) [inline activerecord/lib/active_record/destroy_later.rb:45]: Just had a case for calling this in a controller to defer the record destroy, but I don't want to have it happen `:after` sometime, just out of the current request, so I think we should do:

```suggestion
    def destroy_later(after: nil, ensuring: nil)
```

**Comment by rafaelfranca** [comment]
Closing this PR since we decided to not implement the class level `destroy_later`. `dependent: :destroy_async` was implemented in #40157.
